llama_index/agent/openai_agent.py

```diff
@@ -104,6 +138,66 @@ def chat(
 
         return Response(ai_message.content)
 
+    def stream_chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> Generator[StreamingChatResponse, None, None]:
+        chat_history = chat_history or self._chat_history
+        chat_history.append(ChatMessage(content=message, role="user"))
+        tools = self._get_tools(message)
+        functions = [tool.metadata.to_openai_function() for tool in tools]
+
+        def _stream_chat(
+            chat_history: List[ChatMessage],
+        ) -> Generator[StreamingChatResponse, None, None]:
+            # TODO: Support forced function call
+            chat_stream_response = StreamingChatResponse(
+                self._llm.stream_chat(chat_history, functions=functions)
+            )
+
+            # Get the response in a separate thread so we can yield the response
+            thread = Thread(
+                target=chat_stream_response.write_response_to_history,
+                args=(chat_history,),
+            )
+            thread.start()
+            yield chat_stream_response
+            thread.join()
+
+            n_function_calls = 0
+            function_call = chat_history[-1].additional_kwargs.get(
+                "function_call", None
+            )
+            while function_call is not None:
```

Comment:
is there a way we can share code between `chat`, `stream_chat`, and the async versions? a bit worried about maintaining all 4 versions of the agent loop

Reply:
ha I had the same thought. There's a similar problem across nearly all of our async functions.

I'll see if I can consolidate a bit